[HLSL][NFC] Remove RegisterBindingFlags struct

hekota opened the pull request:
When diagnosing register bindings we just need to make sure there is a resource that matches the provided register type. We can emit the diagnostics right away instead of collecting flags in the RegisterBindingFlags struct. That also enables early exit when scanning user defined types because we can return as soon as we find a matching resource for the given register type.

hekota committed:
[HLSL][NFC] Remove RegisterBindingFlags struct

When diagnosing register bindings we just need to make sure there is
a resource that matches the provided register type. We can emit the
diagnostics right away instead of collecting flags in the
RegisterBindingFlags struct. That also enables early exit when scanning
user defined types because we can return as soon as we find a matching
resource for the given register type.

hekota committed:
revert comment change

hekota committed:
clang-format

hekota commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -41,6 +41,47 @@
 
 using namespace clang;
 
+enum class RegisterType { SRV, UAV, CBuffer, Sampler, C, I, Invalid };
```
This is moved from below.

llvm-beanz commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -828,101 +831,85 @@ static void updateResourceClassFlagsFromRecordType(RegisterBindingFlags &Flags,
   }
   // Samplers, UAVs, and SRVs are VarDecl types
-  else if (VarDecl *TheVarDecl = dyn_cast<VarDecl>(TheDecl)) {
+  if (VarDecl *TheVarDecl = dyn_cast<VarDecl>(D)) {
```
nit: Rather than an unreachable at the end of this block, we could eliminate this block and have this be a `cast` with an `assert(isa<VarDecl>(D)...)` above. That reduces the control flow nesting.

llvm-beanz reviewed (approved):
One minor nit inline, but I think this is fine either way.

damyanp reviewed (commented):
LGTM - just a couple of nits. 

This refactor is leaning heavily on the tests, it might be good to get a review from @bob80905 just in case there's something that the tests didn't cover that may need to be considered.

hekota commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -828,101 +831,85 @@ static void updateResourceClassFlagsFromRecordType(RegisterBindingFlags &Flags,
-                              ? llvm::dxil::ResourceClass::CBuffer
-                              : llvm::dxil::ResourceClass::SRV;
+  if (HLSLBufferDecl *CBufferOrTBuffer = dyn_cast<HLSLBufferDecl>(D)) {
+    llvm::dxil::ResourceClass RC = CBufferOrTBuffer->isCBuffer()
```
Yeah, we should do that.

hekota committed:
Code review feedback

+ couple of local variables names shortened for easier debugging when manually invoking dump() in debugger

hekota committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into inline-res-binding-diag

hekota committed:
clang-format

hekota committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into inline-res-binding-diag

bogner commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -839,100 +842,84 @@ static void updateResourceClassFlagsFromRecordType(RegisterBindingFlags &Flags,
-    }
-  } else {
-    llvm_unreachable("expected be VarDecl or HLSLBufferDecl");
+  assert(isa<VarDecl>(D) && "D is expected to be VarDecl or HLSLBufferDecl");
+  VarDecl *VD = cast<VarDecl>(D);
```
FWIW `cast<>` will assert if the cast fails. It's debatable whether or not the slight amount of extra information from the specific message is worth its own assert, so this doesn't necessarily need to be changed, but I figured I'd point that out.
